### ISSUE TITLE
Making 'three' tree-shakeable

### ISSUE THREAD

**Issue by pschroen** [mrdoob/three.js#24199]
The goal with tree-shaking here is I should be able to import a class from `'three'`, and only get the classes needed. In my test bundle importing only `Vector2` is producing a 295 KB (uncompressed) file still with lots of remaining side-effects even after `r141` and all the work done on #24006 (down from a 1 MB bundle in `r140`).

I'm opening this issue to resolve the remaining side-effects, which is do-able with some more work, and we have a couple ways of testing that.

Also to make the claim that `'three'` is finally fully tree-shakeable, we'll need to verify that in the most popular bundlers. I'll start with [Rollup](https://rollupjs.org/), [Webpack](https://webpack.js.org/), [Parcel](https://parceljs.org/) and [esbuild](https://esbuild.github.io/). Open to suggestions of other bundlers and configurations, and will start with a simple test of importing only `Vector2`.

```js
import { Vector2 } from 'three';

console.log(new Vector2(0, 1));
```

And with [agadoo](https://github.com/rich-harris/agadoo):

```sh
npx agadoo
...
Failed to tree-shake build/three.module.js
```

It's worth noting that importing from the source files with `agadoo` also fails, and is something I can look into as well.

```sh
npx agadoo src/Three.js
...
Failed to tree-shake src/Three.js
```

The expected behavior, regardless of `agadoo`, is simply looking at the output bundle. If I import `Vector2`, I'm expecting only the `@license` header comment, and `Vector2`, nothing else.

References:
 - https://twitter.com/MarcoGomezGT/status/1520866649485754372
 - mrdoob/three.js#24006

**Comment by marcofugaro** [comment]
Other related discussions: #21667 #21665

**Comment by LeviPesin** [comment]
Does `instanceof` work with tree-shaking?
I.e. if, for example, WebGLRenderer has `.isVector3` check in its body but does not have Vector3 import, does the changing it to `instaceof Vector3` and importing Vector3 will break tree-shaking or not? I.e. will the Vector3 class be included?
Related: https://github.com/mrdoob/three.js/issues/24006#issuecomment-1121315138

**Comment by pschroen** [comment]
Oh I think I get why this might be an issue, is the problem simply that `PointsMaterial` would be included in the bundle?

Technically it is being used in that case, and would be included in the bundle. If the goal is to not include any of the materials with `WebGLRenderer`, then ya we would have to use `.isPointsMaterial`.

I'll add this to my list of things to investigate...

**Comment by LeviPesin** [comment]
> Oh I think I get why this might be an issue, is the problem simply that PointsMaterial would be included in the bundle?

Yes.

> If the goal is to not include any of the materials with WebGLRenderer, then ya we would have to use .isPointsMaterial.

I think we should calculate how often such pattern is used (the class is not imported, but `.isClass` is used)...

**Comment by marcofugaro** [comment]
> I.e. if, for example, WebGLRenderer has .isPointsMaterial check in its body but does not have PointsMaterial import, does the changing it to instanceof PointsMaterial and importing PointsMaterial will break tree-shaking or not? I.e. will the PointsMaterial class be included?

Yup, it will be included. This is the reason three switched to `.is*` in #9310.

**Comment by marcofugaro** [comment]
> BTW (completely unrelated to this issue but related to that), is the reasons for not using default exports valid now? They are very useful to avoid constantly writing import { Abc } and export { Abc } (they are already used in e.g. the nodes system).

Code style I believe.

**Comment by LeviPesin** [comment]
> I think we can use the same pattern in most such occurences?

E.g. the first occurence of `.is` that GitHub search shows - https://github.com/mrdoob/three.js/blob/dev/src/renderers/shaders/UniformsUtils.js - can be easily fixed by changing the `value.isVector2 || ...` to `value.clone !== undefined`. Will file a quick PR for this.

**Comment by pschroen** [comment]
@donmccurdy Yes agreed, there's been some confusion in the terminology, when I read that `instanceof` doesn't work with tree-shaking I was taking it literally, technically it does work, it just works against this library because of the way the types are being checked.

I think there's some good discussion here on alternative approaches as above. I'm going to focus more on removing the remaining side-effects in the meantime.

**Comment by donmccurdy** [comment]
> My suggestion is not to replace .is with not-tree-shakable instanceof - my suggestion is to replace it with polymorphism and duck typing

But why? I guess I don't see any benefit to this — just a lot of work, harder-to-read code, and a hard-to-predict chance of causing classic [performance pitfalls associated with polymorphism](https://stackoverflow.com/questions/37058643/why-do-monomorphic-and-polymorphic-matter-in-javascript). If there is some benefit, let's discuss it in another thread since it doesn't seem to be related to this one?

**Comment by pschroen** [comment]
Agreed, even though related, the bundle being tree-shakeable in its current state doesn't hinge on the type check approach. 👍

**Comment by pschroen** [comment]
Alright, so I've spent some more time tonight with the latest, here's a summary of my findings. 😉
 - `src/Three.js` itself has side-effects at the bottom because of the checks for `__THREE_DEVTOOLS__` and multiple instances.
 - `Three.Legacy.js` is fully tree-shakable now, nice work guys! 🎉
 - The `DataUtils` class has a number of expressions and for loops causing side-effects.
 - The `PropertyBinding.js`, `KeyframeTrack.js`, `Materials.js`, 'Object3D.js', `Texture.js` and `Color.js` files have static and prototype properties that could be converted to class fields. Have we made a decision on the use of class fields?
 - `Material.js` has a side-effect at the bottom with `Material.fromType` and a `// TODO: Behavior added in Materials.js`, can that be removed?
 - `ShaderLib.js`, `UniformsLib.js` and `ShaderChunk.js` are also side-effects, perhaps there's a different approach we could take with defining these shaders?
 - `ObjectLoader.js` creates a big side-effect with the `Geometries` namespace, removing that one line allows for tree-shaking.
 - `WebGLRenderer` of-course is loading a lot of classes including the shader classes with side-effects, which in-turn creates a domino effect importing all the classes from `ShaderLib.js` and `UniformsLib.js`, etc., removing those classes allows for tree-shaking as well.
 - And finally `MathUtils.js` has a side-effect at the top with the `_lut` for loop.

Removing all of the above, shader and renderer classes from `src/Three.js` is allowing me to build a module bundle with `npm run build-module` and then in-turn import from `build/three.module.js` side-effect free.

```sh
npx agadoo
...
Success! build/three.module.js is fully tree-shakeable
```

Still lots of work to do but we're getting there! 😅

**Comment by lukastaegert** [comment]
So while relying on duck typing alone is probably best for tree-shaking, `instanceof` *is* an improvement as it allows for some basic static analysis. I created rollup/rollup#4524 for you, you can try out if this would help.

**Comment by mrdoob** [comment]
@lukastaegert Nice!

Maybe we can try using `instanceof` with `Vector2`, `Vector3` and `Vector4`. That should solve #24167.

@pschroen Would you like to give it a go?

**Comment by pschroen** [comment]
Sure, though I feel like it might be a bigger can of worms for consistency? I am a fan of a hybrid approach for type checks, so maybe `.is` properties for classes used by `WebGLRenderer` to keep the bundle size down, and then `instanceof` for the math classes? Is there any performance impact with `instanceof` compared to `.is`?

For reference I was looking at this file first.
https://github.com/mrdoob/three.js/blob/de17358232c3cc45bb2df0a5a64068cbcdeaa2f8/examples/jsm/nodes/core/NodeUtils.js#L33

**Comment by mrdoob** [comment]
>  Is there any performance impact with `instanceof` compared to `.is`?

I want to think that `instanceof` is faster... 🤞

**Comment by LeviPesin** [comment]
So now (after merging that PR, actually) when Rollup tree-shakes `instanceof`, we can convert _all_ of the `.is` checks to `instanceof`?

**Comment by marcofugaro** [comment]
> when Rollup tree-shakes instanceof, we can convert all of the .is checks to instanceof?

No, this would make the whole repo un-treeshakeable again 😅

Rollup isn't the only bundler out there.

**Comment by LeviPesin** [comment]
I think converting some `.is*` to `instanceof` and not converting others would only create a mess.

Maybe we should create few issues for tree-shaking `instanceof` in other popular bundlers?

**Comment by marcofugaro** [comment]
> Maybe we should create an issue for tree-shaking instanceof in other popular bundlers repositories? Or even a PR based on https://github.com/rollup/rollup/pull/4524.

Good luck with that, we tried in the past with tree-shaking `.prototype.` and it was never supported.

**Comment by pschroen** [comment]
Ya agreed, support for Rollup has always been amazing (thanks @lukastaegert ❤️), we can do so much more with tree-shaking in Rollup now, though the goal of this issue is to get tree-shaking working for all major bundlers. 🥲

I do agree with @donmccurdy this should also be a separate issue and discussion now at this point, and I'm on the fence about the consistency of the change, I do like the idea of a hybrid approach with the math classes, however it does also add a level of inconsistency.

Ultimately it's up to you guys, should we try a hybrid approach in a new issue, or stick with the `.is*` type checks for now?

**Comment by LeviPesin** [comment]
> Material.js has a side-effect at the bottom with Material.fromType and a // TODO: Behavior added in Materials.js, can that be removed?

See https://github.com/mrdoob/three.js/pull/24094.

**Comment by pschroen** [comment]
A note on #24218 above,

The advantage of using singleton patterns in general is it reduces the expressions evaluated by the browser at runtime, loading resources only when needed.

Many of these side-effect expressions aren't even being used, but the browser is executing this code regardless, which is why they are side-effects.

Ideally for a library to be "pure" there should be no code running at all outside of the classes. To quote Rich Harris's [agadoo](https://github.com/rich-harris/agadoo)'s again:
> Don't create instances of things on initial evaluation — instantiate lazily, when the functions you export are called

I'd imagine this would improve load time, and also reduce memory usage a little as well. 😉

**Comment by pschroen** [comment]
Referencing @marcofugaro's https://github.com/mrdoob/three.js/pull/24218#issuecomment-1151224477,

Sure, I could convert that to plain functions like in `MathUtils`, that's also one of the recommended approaches from [agadoo](https://github.com/rich-harris/agadoo):

> Export plain functions

The only catch, similar to `MathUtils`'s `_lut` is there's code being evaluated to generate the lookup tables. I suppose I could convert that to an internal `_generateTables()` function which should work with tree-shaking too?

**Comment by pschroen** [comment]
@marcofugaro Is there any reason to keep the `DataUtils` namespace in this case? In other words, is anyone externally using this class, it's in the docs:

https://threejs.org/docs/#api/en/extras/DataUtils

**Comment by marcofugaro** [comment]
> In other words, is anyone externally using this class, it's in the docs:

Yes I believe people are using that class. The way they import would change, like it did for BufferGeometryUtils: https://github.com/mrdoob/three.js/pull/22267

**Comment by marcofugaro** [comment]
Btw you can test tree-shaking by running `npm run test-treeshake`

**Comment by pschroen** [comment]
> Have we made a decision on the use of class fields?

Following-up on this one, I know we've already made a lot of progress since `r140` with removing `.prototype.*` properties, though we still have quite a few static properties that could be converted to class fields. See https://github.com/mrdoob/three.js/issues/24006#issuecomment-1123779387

What's the verdict on this one?

**Comment by LeviPesin** [comment]
I think we don't yet use class fields, optional chaining, and nullish coalescing in the core.
It's maybe a time to change that?

**Comment by pschroen** [comment]
Alright, so @marcofugaro I've created a draft PR above for discussion on using static class fields.

> We moved away from the "class with only static methods" pattern

Note that adding the pure annotations to `ShaderLib` and `UniformsLib` does appear to be working in Rollup with the original object literal as well, so really the only advantage with the static class fields is having the `physical` property inside the class which works with `this`.

Also some great news, because it appears that those two files were the linchpin for tree-shaking the entire library with the latest version of Rollup! 🎉

`ShaderLib` and `UniformsLib` were creating a cascading side-effect across everything that touches them, the shader and renderer classes appear to be fine now, and simply adding the pure annotations to those two files allows for tree-shaking the entire library (with the exception of `src/Three.js` itself, but I'm saving my comments on that until everything else is done). 😊

**Comment by LeviPesin** [comment]
I actually think in this case it is better to have just an object with properties (and add `physical` there, obviously) rather than a class...

**Comment by pschroen** [comment]
@LeviPesin Sure, I'll push a new commit to the draft switching back to the original object literal, and then we can revisit class fields at a later time if needed. 👌

**Comment by pschroen** [comment]
> Btw you can test tree-shaking by running npm run test-treeshake

@marcofugaro I finally got this command to work by downgrading `chalk` to `v4.1.2`, must be something wrong with the `v5` ESM version. I'm running the latest version of Node `v18.3.0`.

Not sure about this as a tree-shaking test though, I would much prefer [agadoo](https://github.com/rich-harris/agadoo) if we can get it working (it's currently broken as well because it's pre-Rollup `v2`).

Investigating a solution for either or, ideally we'd want something that will error-out we can add to three's `test` script. It's super easy to accidentally add side-effects in a library and think a goal of this issue should also be a success/fail unit test. 😉

**Comment by pschroen** [comment]
I should also clarify something here:

> tree-shaking the entire library

I've edited that to be "flat bundle" instead, really this is tree-shaking a single class without side-effects at the highest level (which is why I chose `Vector2` to start).

But there are plenty of deeper level tree-shaking issues, I would consider those secondary issues we can also solve with real world examples. A simple practical example is also easy to write a unit test for, for example:

```js
import { BoxGeometry } from 'three';

console.log(new BoxGeometry());
```

```js
import { MeshBasicMaterial } from 'three';

console.log(new MeshBasicMaterial());
```

```js
import { MathUtils } from 'three';

console.log(MathUtils.clamp(0, 0, 1));
```

```js
import { clamp } from 'three';

console.log(clamp(0, 0, 1));
```

I'll create separate issues for these, though for this issue I would consider tree-shaking `Vector2` and a high-level unit test like [agadoo](https://github.com/Rich-Harris/agadoo) enough to close this particular issue.

One last note on the terminology of "tree-shaking" that has been confusing a lot of people, and why I don't view tree-shaking as dead code removal, it's really the opposite, [live code inclusion](https://en.wikipedia.org/wiki/Tree_shaking).

So I'm more thinking about what it's including, it's the opposite approach, like I shake this tree here and out drops only `Vector2`. 😉

**Comment by pschroen** [comment]
I've created a draft PR above for discussion on the registration event and multiple instances check, and if it's really needed on initial evaluation, or if it's fine when `WebGLRenderer` is instantiated? How would this work with the other renderers as well?

**Comment by pschroen** [comment]
Referencing @gkjohnson's https://github.com/mrdoob/three.js/pull/24225#issuecomment-1153069609 cc @Mugen87

Hmm, good points on both the **"one library, multiple renderers"** and **"two libraries, one renderer"**. This is the reason why I was leaving this conversation till the end of troubleshooting this issue (for Rollup anyways).

For the second point, **"two libraries, one renderer"**, this really conflicts with tree-shaking in general which needs a pure scope with nothing on initial evaluation. If the only way **"two libraries, one renderer"** can be checked is with something in global scope, on initial evaluation, then unfortunately this goes back to my use case points on how people use the library. See https://github.com/mrdoob/three.js/issues/24006#issuecomment-1120203227 😜

Let me revise it then, so in my opinion there are three use cases:

- A **"THREE namespace, script"**, like in the manual or when loading the pre-bundled library from a CDN, the multiple instances check makes sense here.
- A **"THREE namespace, ES module import"**, like in the docs or examples, the multiple instances check makes sense here too, though I feel like the onus is more on the application developer with this one.
- An **"ES module library"**, so I'm potentially not using a renderer and importing utility type classes, this is also the use case where more and more people are importing directly from the source files to get around the tree-shaking problems.

So given that, we could solve this particular problem if we were exporting the source files as the library entry point, keeping the bundles, you could just move the registration event and multiple instances check to the Rollup builds, appended to the bottom of the bundles, and then for people importing from the source files it wouldn't be a problem for them.

So unless we can think of an alternative approach, I think we're stuck in a little bit of a "catch-22" type of situation here, to fully tree-shake this library we'll need to remove that functionality. I realize it's there for convenience for new developers, but really I would argue loading two instances of the library, or accidentally bundling two namespaces (double the bundle size) isn't the responsibility of a library, the onus is on the application developer.

You could make this argument with any library. Perhaps we replace it with a "best practices" kind of thing in the docs or manual on how developers can avoid these accidental situations? This feels more like a developer problem imho. 😉

**Comment by gkjohnson** [comment]
> For the second point, "two libraries, one renderer", this really conflicts with tree-shaking in general which needs a pure scope with nothing on initial evaluation.

I haven't been following the full thread abut where has it been shown that these lines need to be removed in order to tree shake the full library?

That aside I think there are still improvements that can be made to that section of code:

- I think the `__THREE_DEVTOOLS__` event dispatch can be safely moved to the WebGLRenderer initialization as long as it's only run once which means we only have to worry about the multiple imports check.
- As you suggest move the multiple instances checks to just the bundled code which will cover the new-user use cases and catch issues there. And if you want advanced, full-tree-shaking behavior you can configure your bundler to pull from "src/Three.js"

> really I would argue loading two instances of the library, or accidentally bundling two namespaces (double the bundle size) isn't the responsibility of a library, the onus is on the application developer

Unfortunately we're still seeing cases where people are running into the "multiple instances of three" problem on the forum so I don't think it should go if we can avoid it - it's a really difficult problem to debug without with warning.

**Comment by LeviPesin** [comment]
> As you suggest move the multiple instances checks to just the bundled code

I support this idea.

**Comment by pschroen** [comment]
Sounds good, I'll update the draft moving them to the bundled code.

@gkjohnson Ya these are causing side-effects, adding pure annotations to these also makes no difference with tree-shaking because they are global scope. I'll think on other possible ways around this as well...

**Comment by Mugen87** [comment]
> I'd be perfectly happy to see src/ removed from the npm package.

Yes, that sounds reasonable. I would also prefer to just have single way how classes from the core are imported 👍 .

However, I do not vote for having more than one `three` npm package. That would make the rollout and maintenance just more complicated.  

I'd also like to raise the question if the project needs to be perfectly tree-shakable. 100% solutions are often tricky to implement and the added value not always justifies the efforts. Maybe we can just state that the current result is "good enough" and focus on other topics?

**Comment by donmccurdy** [comment]
> In other words, I am in support of removing src/ from the npm package, so long as we give users another way to access the source files via npm.

I don't think it is a good idea to mix npm and grabbing src/ files, or to have multiple versions of the library on npm. If users prefer to manage dependencies manually, without NPM, they can do so quickly:

```
npx degit mrdoob/three.js/src#r141  ./three-src
```

Still, this is not a workflow I would encourage anyone to use unless you're _very_ sure of what you're doing and its implications. Personally I've stopped responding to forum questions questions related to project setup, unless the author is using npm and a bundler. Providing support for so many installation methods is an endless waste of everyone's time.

**Comment by pschroen** [comment]
@Mugen87 I hear ya, maybe we just leave this as is then for now, as @gkjohnson put it above:

> And if you want advanced, full-tree-shaking behavior you can configure your bundler to pull from "src/Three.js"

That's what started this whole conversation with @MarcoGomezGT's [tweet](https://twitter.com/MarcoGomezGT/status/1520866649485754372) on the topic. He's not the only one doing this, most developers I'm working with these days are all working this way already, myself included.

I don't normally stir the pot on these things though after years of dealing with this same problem myself I wanted to clear-up some confusion on it, and wanted to give it a go myself. I'm pretty happy with the results we've achieved in Rollup, but there's more that can be done with the other bundlers.

**Comment by gkjohnson** [comment]
> I'd also like to raise the question if the project needs to be perfectly tree-shakable. 100% solutions are often tricky to implement and the added value not always justifies the efforts. Maybe we can just state that the current result is "good enough" and focus on other topics?

I agree - it would be nice to quantify the effect of the multiple instances check a bit more, though. How much is it impacting the tree shake-ability?

Assuming it's significant I think there may still be some alternatives to consider before giving up on tree shaking - first as I mentioned in I think the `__THREE_DEVTOOLS__` can safely be moved to the `WebGLRenderer` constructor without any loss in functionality which leaves us with just the multiple instances check.

For the multiple instances check I know at least webpack has the ability to prune code encapsulated in constant conditions, such as when [building for production](https://webpack.js.org/guides/production/#specify-the-mode) (it's an option, at least):

```js
if ( typeof process !== 'undefined' && process.env.NODE_ENV !== 'production' ) {

    // this code will be pruned in production builds

}
```

Maybe we can add a define flag for three or use an existing production flag to enable that multiple instances check to be pruned only in prod builds? Do we know if there's a standard for this kind of thing across bundlers?

**Comment by pschroen** [comment]
@marcofugaro Yes there's that too, good point! 😄

**Comment by LeviPesin** [comment]
> even with all of @lukastaegert's latest updates

They were closed, unfortunately :-(

**Comment by donmccurdy** [comment]
I see https://github.com/rollup/rollup/pull/4524 (tree-shaking `instanceof`) didn't go in – were there others closed?

**Comment by LeviPesin** [comment]
> I see https://github.com/rollup/rollup/pull/4524 (tree-shaking instanceof) didn't go in – were there others closed?

I was talking only about it, were there other related changes?

> Is that specific code not being tree-shaken (okay), or is it preventing other things from being tree-shaken (less okay)?

The first (and, also, multiple instance checking).

**Comment by donmccurdy** [comment]
Ah ok! If it's just the `instanceof` change that didn't go through, and we still have our `.isFoo` approach, so that's manageable.  👍🏻

**Comment by pschroen** [comment]
> Since latest release, importing from 'three' produces the same result as importing from 'three/src/Three', so there is really no point in importing from src/ anymore.

Hey so @marcofugaro, I've moved over to webpack now and if we're talking about the latest release of `r141` that's not the case. The only difference in these two logs is importing from `'three/src/Three'`.

```js
import { Vector2 } from 'three';

console.log(new Vector2(0, 1));
```

```sh
npx webpack
asset bundle.js 1.21 MiB [compared for emit] (name: main)
runtime modules 396 bytes 2 modules
cacheable modules 1.09 MiB
  ./src/index.js 66 bytes [built] [code generated]
  ./node_modules/three/build/three.module.js 1.09 MiB [built] [code generated]
webpack 5.73.0 compiled successfully in 295 ms
```

```js
import { Vector2 } from 'three/src/Three';

console.log(new Vector2(0, 1));
```

```sh
npx webpack
asset bundle.js 10.4 KiB [emitted] (name: main)
orphan modules 1.16 MiB [orphan] 366 modules
runtime modules 396 bytes 2 modules
cacheable modules 5.88 KiB
  ./src/index.js 76 bytes [built] [code generated]
  ./node_modules/three/src/math/Vector2.js 5.8 KiB [built] [code generated]
webpack 5.73.0 compiled successfully in 347 ms
```

**Comment by MarcoGomezGT** [comment]
`The thing that that tweet got wrong, is that contributors went through a lot of work to allow tree shaking to work for the 'three' import, not three/src/Three sweat_smile`

@marcofugaro  I wrote that tweet, and when I mentioned the work of the contributors I was referring to the substantial work dedicated to the conversion to ES6 classes, and what I meant (explaining my - dirty, I must admit 😅, but effective - hack further in the same [thread](https://twitter.com/MarcoGomezGT/status/1521096343904739330))  was that even with all the progress achieved, tree-shaking is still not possible with Rollup or Webpack so far (as demonstrated above by @pschroen with tests that I can back up and confirm on my end), unless you're willing to use the suggested hack (which can cause severe issues as any newcomer importing some other lib that imports from 'three' (the ideal way) will end up with an even bigger bundle and the lib imported twice under different namespaces (build and src).

The seeds were planted, and we're almost there, but we're just not harvesting the fruits yet (at least not the way we're supposed to).

**Comment by marcofugaro** [comment]
Yeah, not true at all. Here is the correct size difference.

```js
import * as THREE from 'three'

const camera = new THREE.PerspectiveCamera( 70, window.innerWidth / window.innerHeight, 0.01, 10 );

const scene = new THREE.Scene();

const renderer = new THREE.WebGLRenderer( { antialias: true } );
renderer.setSize( window.innerWidth, window.innerHeight );
renderer.setAnimationLoop( animation );
document.body.appendChild( renderer.domElement );

function animation( time ) {

	renderer.render( scene, camera );

}
```
<img width="293" alt="image" src="https://user-images.githubusercontent.com/7217420/173512757-4515c9b6-2ac9-4723-84fa-2e73e3ab31e1.png">

versus

```js
import * as THREE from 'three/src/Three.js'

const camera = new THREE.PerspectiveCamera( 70, window.innerWidth / window.innerHeight, 0.01, 10 );

const scene = new THREE.Scene();

const renderer = new THREE.WebGLRenderer( { antialias: true } );
renderer.setSize( window.innerWidth, window.innerHeight );
renderer.setAnimationLoop( animation );
document.body.appendChild( renderer.domElement );

function animation( time ) {

	renderer.render( scene, camera );

}
```

<img width="291" alt="image" src="https://user-images.githubusercontent.com/7217420/173512919-db6f83ba-09a4-4a32-a0f4-5a5bcdc1fcde.png">


As you can see the size difference is minimal. You can reproduce this test [on this repo I made](https://github.com/marcofugaro/treeshake-test).

------------------

Webpack has no support for tree-shaking static class properties, so until we switch to the `static` keyword, the WebGLRenderer will not be tree-shakeable. But it is not important right now, since WebGLRenderer is used almost always.


> I mentioned the work of the contributors I was referring to the substantial work dedicated to the conversion to ES6 classes,

Yeah it was done in order to make tree-shaking work while importing from `'three'`. three.js never encouraged importing from `three/src`.

**Comment by MarcoGomezGT** [comment]
@marcofugaro Again, confirming what @pschroen kindly posted before...

```
import { Vector2 } from 'three'
const v2 = new Vector2(1, 2)
console.log(v2)

```
![Screenshot from 2022-06-14 08-45-00](https://user-images.githubusercontent.com/33723163/173522194-8b9a02fe-c2e8-43cb-b7e1-5c3e525eb541.png)

while:

```
import { Vector2 } from 'three/src/math/Vector2'
const v2 = new Vector2(1, 2)
console.log(v2)
```

![Screenshot from 2022-06-14 08-46-10](https://user-images.githubusercontent.com/33723163/173522548-3bbdc1de-c80e-485d-868a-0033760094e2.png)

Are we missing something you're not?

**Comment by marcofugaro** [comment]
Again, 

> Webpack has no support for tree-shaking static class properties, so until we switch to the static keyword, the `WebGLRenderer` will not be tree-shakeable on webpack. But it is not priority right now, since `WebGLRenderer` is used almost always.

**Comment by MarcoGomezGT** [comment]
@marcofugaro  I fail to understand why you're mentioning `WebGLRenderer` when we both gave you an example of a half-Megabyte bundle (probably enough for a demoscene to present the path-traced world's history directed by J.J. Abrams) that is solely importing Vector2. I apologize in advance, I'm not trying to sound intransigent (I'm just trying to better illustrate my - perhaps wrong - perspective).

What we're saying doesn't go beyond "Three.js has an issue with tree-shaking through any bundler", and it has. That's perfectly fine and I'm not implying that should be among the project's priorities despite I believe it should. All I did was present a hack to my followers, that I personally use on my personal projects as a workaround for the issue we're talking about (that exists, and was demonstrated in several different ways since the start of this conversation).

**Comment by lukastaegert** [comment]
While I am following this discussion with one eye, one data point I want to add is why it might be important to be able to tree-shake the case where only "Vector2" is used even though one would expect that it does not make sense to use Three.js without using the renderer: code-splitting.
Imagine a situation where Three.js itself is loaded lazily, but you want to do some pre-calculations beforehand using e.g. the vector class. If the case where only the vector class is used cannot be tree-shaken properly, then most of Three.js will go into the initial chunk. On the other hand if it can be tree-shaken, then lazy loading actually makes sense and the initial chunk will not carry the burden of the entire renderer. While this is definitely true for Rollup, I would expect that it is also true for every other bundler that does both code-splitting and tree-shaking.

**Comment by marcofugaro** [comment]
@MarcoGomezGT Ok, let me try explain it to you more clearly. I do not mean to be stern in any way.

These are the fundamental three.js modules needed to run any basic scene. These are all tied to the `WebGLRenderer`. You have to use all of these if you want to spin up any project in three.js, including the [basic scene illustrated in the README](https://github.com/mrdoob/three.js#usage).

**These modules weight ~420kb. You can't get a three.js scene running which weights less, not even importing from `src/`.**

<img width="1359" alt="image" src="https://user-images.githubusercontent.com/7217420/173530768-3f1b8967-5fcb-452f-999a-49b232a5b0d3.png">

As you can see, not every module from the `three/src` folder is including, for example the [animation modules](https://github.com/mrdoob/three.js/tree/dev/src/animation). This is tree-shaking effectively working.

As as result of this, if you switch a **real-working three.js demo** from importing from `'three'` to `'three/src/Three'` (or vice-versa) the size difference is almost nothing.

--------------------------

It is not the same if you import just `Vector2` on webpack. Since webpack has issue with tree-shaking the fundamental three.js modules (screenshot above), it includes them anyway.

But since importing only `Vector2` is not a real-world scenario, it is not really that priority right now.

-------------------------

Hope that clears things up, let me know if you have any other questions.

**Comment by MarcoGomezGT** [comment]
@marcofugaro I completely understand. Unfortunately, I couldn't find the time to follow up on the discussion closely, so I'm not aware if something changed in `r141`. What I was trying to say before is that I could quickly confirm that the findings presented by @pschroen were correct when he stated that importing from 'three' and from 'src' still produces different results on the test scenario in which he just imported Vector2.

What I understand you're telling me is that, after r141, the basic scene from the readme will produce the same bundle size through Webpack regardless of importing from 'three' or from 'src', is that correct?

Also, I'm under the impression that @lukastaegert presented a perspective that justifies being able to import less intricate and smaller classes beforehand in lazy-loading scenarios, and his argument seems reasonable. Example: pre-computing at some point a Fibonacci sphere of a given (possibly big) amount of samples and a scale just making usage of Vector3 in advance, before loading the rest of the library.

```
import { Vector3 } from 'three/src/math/Vector3'

export const fibonacciSphereV3 = (samples, scale) => {
  const gAng = Math.PI * (3.0 - Math.sqrt(5.0))
  const points = []
  for (let i = 0; i < samples; i++) {
    const y = 1.0 - (i / (samples - 1)) * 2.0
    const radius = Math.sqrt(1.0 - y * y)
    const theta = gAng * i
    const x = Math.cos(theta) * radius * scale
    const z = Math.sin(theta) * radius * scale
    const v3 = new Vector3(x, y * scale, z)
    points.push(v3)
  }
  return points
}
```
Anyway... Just to clarify about the beginning of my thread on Twitter, please find below the bundle of a real-world scenario (a complete scene) imported `from 'three'`, and imported `from 'src...'`. IIRC, I was using r140 in such a comparison.

Imported `from 'three'`:

![from_three](https://user-images.githubusercontent.com/33723163/173541017-c6515da6-2a33-4834-90a9-eb9d5bd9c1c2.jpeg)

Imported `from 'src'`:

![from_src](https://user-images.githubusercontent.com/33723163/173541078-a4bf0546-955b-4750-a68b-897f0b779430.jpeg)

As you may notice, when I suggested the hack as a possible workaround for such a difference in bundle size, the usage of such a hack was absolutely justified (probably necessary in some cases where bundle size is a valid concern) as the entire lib was being bundled unnecessarily.

I'll probably find the time soon to make new tests on a real-world scenario to stay informed about the progress, but I appreciate this discussion anyway.

**Comment by gkjohnson** [comment]
Not to continue to stir the pot but I'll share a maybe less contrived use case where three is used without a WebGLRenderer but is not tree-shaken effectively.

In [three-mesh-bvh](https://github.com/gkjohnson/three-mesh-bvh) the user can run BVH generation in a WebWorker to build the structure in a separate thread while the main thread continues to render _with_ a WebGLRenderer. Ideally code splitting could be used here to minimize the number of files loaded or at least only a minimal number of classes required (math, BufferGeometry, BufferAttributes classes).

At this point I may just be repeating the above but for some reason when using rollup and just importing `MeshBVH` imports all the Materials. When using [@rollup/alias](https://www.npmjs.com/package/@rollup/plugin-alias) and aliasing `three` to `three/src/Three.js` the materials are excluded.

However after some experimenting I've found that just importing `EventDispatcher` from `three` _also_ includes all the materials. I'll say that at some point I think it has to be a failing of the bundlers to properly tree shake these "flat bundles" and maybe some issues should be created in those repos to figure out what's going on and what is / isn't considered "three-shakeable".

**Comment by pschroen** [comment]
Hey @gkjohnson, I don't mind stirring the pot on this topic either, it'll be good to get to the bottom of it! 😉

Here's another way of thinking of it, and @lukastaegert please correct me if I'm wrong here, my understanding is all bundlers handle parsing source files differently from a flat bundle simply because of scope.

Let's not forget that bundlers are literally designed to bundle source files, and there are quite a few optimizations and benefits over a flat bundle. For example if only importing `MeshBVH` from your source file entry point, then only the `MeshBVH.js` file is imported with its self-contained scope. Compared to a flat bundle, where the entire file needs to be parsed as if it's being interpreted by the browser.

In other words, flat bundles are riddled with side-effects and require pure annotations to help combat that, while importing from source files is a lot cleaner with isolated scopes. So "flat bundles" are not the same as source files, and never has been. I'm not sure how this misconception has started, but that's my experience anyways. 😅

**Comment by lukastaegert** [comment]
Well, that is actually not true, at least not for rollup. For vanilla rollup, there is not a lot of difference between a flat bundle and importing the entry point of an equivalent file structure. The tree-shaking algorithm does not even know about module boundaries. There is only a difference if `moduleSideEffects` are disabled (this would be equivalent to the `sideEffects` flag in `package.json`). The effect of this is that basically each module is treated as if wrapped in a big pure annotation. The story is slightly different for Webpack, but they delegate the actual tree-shaking to terser anyway, which again works on the generated output + some annotations that Webpack adds based on the module graph.
So my guess is that you heavily rely on these "implicit annotations" via the sideEffects property. The logic here is that a module is include only if at least one export of the module is included, at which point the regular side effect detection logic would kick in for that module.

**Comment by marcofugaro** [comment]
@gkjohnson good point! I think we can start looking into static class properties for this particular usecase.

BTW if you have a repro of the "importing all materials" issue I can tell you why it's not working.

**Comment by pschroen** [comment]
Got it, thanks for confirming! 🙏

That makes sense and also explains the behaviour we're seeing. I'm not sure if this is really a bundler problem though, I see this as the expected behaviour. 🤷‍♂️

**Comment by gkjohnson** [comment]
@lukastaegert @pschroen 
I guess I find it confusing as to why sometimes rollup can treeshake classes that are from a single file and sometimes not. In the case of importing `EventDispatcher` from `three` there are no references to other classes in the `EventDispatcher` class - it would seem to follow that that would be a case where all other classes could be removed from the build. Same goes for `Vector2`. I know the problem might be harder with a flat bundle / have fewer hints for tree shaking but as someone who's not an expert in tree shaking the current behavior doesn't seem consistent so I'm trying to understand when and why these issues occur.

@marcofugaro 
As a simple case the following code winds up including all the material information in the final bundle:

```js
import { EventDispatcher } from 'three';

console.log( EventDispatcher );
```

And the rollup config I'm using:

```js
import { nodeResolve } from '@rollup/plugin-node-resolve';
import alias from '@rollup/plugin-alias';

export default [
	{
		input: './test-bundle.js',
		treeshake: true,

		output: {

			format: 'esm',
			file: './test-bundle.output.js',

		},
		plugins: [
			// uncomment to minimize bundle
			// alias({ entries: [{ find: 'three', replacement: 'three/src/Three.js' }] })
 			nodeResolve()
		],

	}
];

```

**Comment by pschroen** [comment]
@gkjohnson, I can confirm my test bundle behaves the same way. The reason why is there are other classes in the flat bundle that extend `EventDispatcher` and are marked as a side-effect, but only in the flat bundle.

`EventDispatcher` is a great test and example because so many classes extend it, just one here for example:

```js
class Texture extends EventDispatcher {
...
```

**Comment by gkjohnson** [comment]
> The reason why is there are other classes in the flat bundle that extend EventDispatcher and are marked as a side-effect, but only in the flat bundle.

I don't follow. Why are these other class definitions being marked as causing / being side effects? `Vector2` causes the same issue but there are no classes that extend it.

Here's a simple repro that shows just extending a class does not cause the child class to be included if the parent class is required:

```js
// classes.js
export class ClassA {}

export class ClassB extends ClassA {}
```

```js
import { ClassA } from './classes.js';

console.log( ClassA );

// ClassB is not included in the bundle when using rollup
```

**Comment by gkjohnson** [comment]
I see. Thanks for the explanation. I see the issue is the way materials are handled now (which has been discussed previously and is a known issue) - specifically this code looks like it would be problematic:

https://github.com/mrdoob/three.js/blob/dev/src/materials/Materials.js

**Comment by pschroen** [comment]
You got it! And there is a domino effect that happens as well, so when `Material` is flagged as a side-effect, everything that file touches also becomes a side-effect and is included in the bundle because the bundler thinks it's all needed (unless it's marked with a pure annotation). 🙄

Here's a Repl you can play around with, parts of this taken from the Rollup [treeshake](https://rollupjs.org/guide/en/#treeshake) docs as well:
https://rollupjs.org/repl/?version=2.75.6&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QiUyMENsYXNzQSUyQyUyMENsYXNzQiUyMCU3RCUyMGZyb20lMjAnLiUyRmNsYXNzZXMuanMnJTNCJTVDbiU1Q24lMkYqJTQwX19QVVJFX18qJTJGJTIwY29uc29sZS5sb2coJ3NpZGUtZWZmZWN0JyklM0IlNUNuJTVDbmNvbnNvbGUubG9nKCUyMENsYXNzQSUyMCklM0IlNUNuJTVDbiUyRiUyRiUyMFRyeSUyMHJlbW92aW5nJTIwdGhpcyUyMHB1cmUlMjBhbm5vdGF0aW9uJTVDbiUyRiolNDBfX1BVUkVfXyolMkYlMjBuZXclMjBDbGFzc0IoKSUzQiU1Q24lMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJjbGFzc2VzLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGNsYXNzJTIwQ2xhc3NBJTIwJTdCJTdEJTVDbiU1Q25leHBvcnQlMjBjbGFzcyUyMENsYXNzQiUyMGV4dGVuZHMlMjBDbGFzc0ElMjAlN0IlNUNuJTIwJTIwY29uc3RydWN0b3IoKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjBjb25zb2xlLmxvZygnc2lkZS1lZmZlY3QnKSUzQiU1Q24lMjAlMjAlN0QlNUNuJTdEJTVDbiUyMiUyQyUyMmlzRW50cnklMjIlM0FmYWxzZSU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE

**Comment by LeviPesin** [comment]
So we just need to merge https://github.com/mrdoob/three.js/pull/24094? (And decide are we going to use static fields, which I strongly support)

**Comment by marcofugaro** [comment]
Ok I have filed #24237, let's see if we can use class properties...

**Comment by pschroen** [comment]
Hey @gkjohnson @marcofugaro,

Spent some time last night putting together a proof of concept that also shows how tree-shaking behaves very differently with source files. The example below is using Rollup, but we've been experiencing similar behaviour in webpack too.

I've setup a GitHub Actions workflow that downloads and modifies the latest version of three.js from npm, and even fully removes all the `/*@__PURE__*/` annotations to make the point I've been talking about that they aren't needed when importing from source files. Likely what I was confusing as scope, as @lukastaegert pointed out, is really that the files are marked pure by default. See https://github.com/mrdoob/three.js/issues/24199#issuecomment-1155648549

> The effect of this is that basically each module is treated as if wrapped in a big pure annotation.

Note that this test is also `r141`, so it doesn't even include out latest updates, and even the registration event and multiple instances check that is still present in `src/Three.js` are not included in the bundle. As you'll see there is clearly a lot of parsing differences between the flat bundle and the source files.

This effectively creates a pure ESM package of three.js, and don't worry @donmccurdy, I'm not publishing it. 😜

I'm more of a command-line user, feel free to test it out any way you like. 😉

https://github.com/pschroen/three-src/blob/main/.github/workflows/main.yml

```sh
mkdir test
cd test
npm init -y
npm i pschroen/three-src rollup @rollup/plugin-node-resolve
echo "import { EventDispatcher } from 'three';console.log(EventDispatcher);" | npx rollup -p node-resolve
```

And I've also created a `license` branch as an example of how you can just add the `@license` comment to the top of the entry point file. This is what other ES module libraries do as well like [GSAP](https://github.com/greensock/GSAP). Note I'm also using the "pinned comments" format of `/*! */` which works better with minifying.

However in my tests neither the pinned comment, registration event and multiple instances check in `src/Three.js` are making it into the bundle when importing from the source files. Just `EventDispatcher.js` is output.

@lukastaegert, do you have any further insight on why importing from source files is working this way? I mean, I'm not complaining, this is the actual behaviour I want, but it would be nice if the `@license` comment was at-least included?

https://github.com/pschroen/three-src/commit/51583352a1b5d4e7bd4bed901892fc23fede5d13

```sh
npm i pschroen/three-src#license
echo "import { EventDispatcher } from 'three';console.log(EventDispatcher);" | npx rollup -p node-resolve
```

**Comment by lukastaegert** [comment]
> but it would be nice if the @license comment was at-least included?

Rollup is not adding comments in files that are not included. The assumption is that the comment describes the file in question. Unfortunately we do not have a special logic for license comments specifically yet. But any comment at the top of a file that is included is also included in the output, which was usually good enough for license comments.

> However in my tests neither the pinned comment, registration event and multiple instances check in src/Three.js are making it into the bundle when importing from the source files

That is because of your use of `sideEffects: false` in `package.json`, Rollup will only look into a file if there is an actual binding imported from that file that is actually used. Or if it is reexporting something that is exported. Just remove this line from `package.json` and watch the dramatic effect this will have on your output.

This does not affect the comment, though, unless there is really something else in the file that has a side effect, for the rationale above.

An in-between way of doing things could also be to provide a code-split bundle where you have as entry points the main index file as well as all files that contain stuff that should be usable separately. Then Rollup will try to bunch stuff into as few chunks as it deems sensible with no code-overhead over the flat bundle, but you can still benefit from the `sideEffects: false` field if you want to.

**Comment by pschroen** [comment]
Got it, ya this makes sense, this is the behaviour we're experiencing with the main three.js package as well.

https://github.com/mrdoob/three.js/blob/46b01b321e27a80d12708201b2b9e1cb734310e0/package.json#L21

So I take it then that this is only being applied to the `src/*` files in the package, but not to the module entry point?

https://github.com/mrdoob/three.js/blob/46b01b321e27a80d12708201b2b9e1cb734310e0/package.json#L7

**Comment by pschroen** [comment]
Right, that's my understanding as well. Thanks for confirming! 🙏

**Comment by pschroen** [comment]
So I've switched back to webpack and it looks like mrdoob/three.js#24225 is a moot point, webpack behaves the same as Rollup.

The pinned comment, registration event and multiple instances check in `src/Three.js` are all ignored anyways when importing from the source files. I don't know if this behaviour has changed recently, though this is the case with the latest versions of these bundlers.

So given that, I'm going to move to close this issue again now. There are some remaining static properties that can be converted to class fields, but that's a larger conversation and think it warrants a separate issue for that?

**Comment by GuskiS** [comment]
Not sure on whats already done, but can https://github.com/mrdoob/three.js/issues/21665#issuecomment-821292769 be looked again? Couple of months when I checked the bundle, shader related stuff was one of the biggest bits of unneeded code in my bundle. It also seems to be somewhat true in picture of https://github.com/mrdoob/three.js/issues/24199#issuecomment-1154903378
I'm no expert in three.js, but are all those shaders needed for blank scene example? 🤔

**Comment by pschroen** [comment]
Hi @GuskiS, currently all the shader chunks are being exported because they are all in one namespace, and also accessed as properties for convenience. For example:

https://github.com/mrdoob/three.js/blob/191a1ef699a5d53de2cd645c667e7fee184a18fd/examples/webgl_shadowmap_pcss.html#L232

https://github.com/mrdoob/three.js/issues/21665#issuecomment-821292769 as @bhouston suggested is likely the only way forward here, it's also the same approach used by @vanruesc in [Post Processing](https://github.com/pmndrs/postprocessing), and works with tree-shaking.

Here the shader chunks are all self-contained in each of the materials, so when tree-shaking you only get the shaders needed with the materials you're importing. For example:

https://github.com/pmndrs/postprocessing/blob/main/src/materials/CopyMaterial.js

The downside to this, is it's common practice in three.js to override these shader chunks with your own code like above, so we'd also have to provide an easy approach for that. There is already a kind of hacky method we've been using like in the modified material example:

https://github.com/mrdoob/three.js/blob/191a1ef699a5d53de2cd645c667e7fee184a18fd/examples/webgl_materials_modified.html#L89

So that could potentially still work because it's already material-based, but we'd have to effectively remove the `ShaderChunk` namespace everywhere else and would be a pretty large breaking change for people using that namespace.

I'll catch-up on these threads to see if there's anything that can be done there. 🤔

**Comment by pschroen** [comment]
Hey @lukastaegert 👋, think I've found another thing to investigate, this time while troubleshooting `EventDispatcher` which is behaving a little differently compared to `Vector2`, due to the number of classes extending it and possibly related to the original static properties issue.

 - rollup/rollup#4489

Here's the Repl:
https://rollupjs.org/repl/?version=2.76.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmNsYXNzJTIwRXZlbnREaXNwYXRjaGVyJTIwJTdCJTdEJTVDbiU1Q25jbGFzcyUyME9iamVjdDNEJTIwZXh0ZW5kcyUyMEV2ZW50RGlzcGF0Y2hlciUyMCU3QiU3RCU1Q24lNUNuJTJGJTJGJTIwVHJ5JTIwcmVtb3ZpbmclMjB0aGlzJTIwc3RhdGljJTIwcHJvcGVydHklNUNuT2JqZWN0M0QuRGVmYXVsdE1hdHJpeEF1dG9VcGRhdGUlMjAlM0QlMjB0cnVlJTNCJTVDbiU1Q24lMkYlMkYlMjBUcnklMjByZW1vdmluZyUyMGJvdGglMjB0aGUlMjBwdXJlJTIwYW5ub3RhdGlvbiUyMGFuZCUyMGNvbnNvbGUubG9nKCklNUNuJTJGKiU0MF9fUFVSRV9fKiUyRiUyMGNvbnNvbGUubG9nKEV2ZW50RGlzcGF0Y2hlciklM0IlNUNuJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==

**Comment by marcofugaro** [comment]
Guys, I think this issue can be closed for now.

**Comment by pschroen** [comment]
Alright, so @marcofugaro I've spent some more time with webpack tonight as there are still further optimizations we can do with `r142` that are on-topic for this issue, here's a summary of my findings again. 😉
 - As we know there's a large amount of side-effects from all the `.prototype.*` properties, especially with the `animation` classes that are always bundled regardless.
 - `AnimationUtils` can be converted to a plain functions export like we did with `DataUtils`.
 - The `Interpolant` classes are also always included too because of the `animation` classes (domino effect).
 - The `ShaderChunk` object literal as we know is a big one, @marcofugaro I know you're moving to close this issue, though for webpack at-least this is still on-topic with tree-shaking `Vector2`, you noted [#21665](https://github.com/mrdoob/three.js/issues/21665) as the first comment of this issue and even recommended a solution there. See https://github.com/mrdoob/three.js/issues/21665#issuecomment-821510530

Also had a question on all the `WebGL*` files, is there a particular reason those have been left as revealing module patterns instead of being converted to ES6 classes?

**Comment by LeviPesin** [comment]
ES6 classes are just cleaner and easier to read.

**Comment by Mugen87** [comment]
> This is a can of worms question, are there any other ways of defining a WebGL 1 rendering context, perhaps by passing-in a parameter to the constructor?

You can pass in a custom rendering context to `WebGLRenderer`.

**Comment by pschroen** [comment]
Ya I'm not sure if deprecating the `WebGL1Renderer` class is worth the 2 lines of code that gets bundled. 😆

**Comment by Mugen87** [comment]
No, I don't think so^^.

**Comment by pschroen** [comment]
As an update on tree-shaking the shaders, I've spent some time yesterday and today trying to make them fully tree-shakeable, I've almost got it working, but hit a couple blockers preventing me from finishing the work.

1. It's especially difficult making three.js tree-shakeable with the flat bundle, and almost contradictory exporting all the shaders so they can be overridden, while at the same time tree-shaking them all when not used.

2. When converting the contents of `ShaderLib` and `UniformsLib` to either named exports, or even individual files, Rollup always bundles the namespaces in the flat bundle as frozen objects with `Object.freeze()`, so even though I can get this working with tree-shaking it then breaks some examples and `RectAreaLight` because of the way we're using them as writable objects for custom shaders, for example:

https://github.com/mrdoob/three.js/blob/cf25a80737db2914f0b36e7c77af23a9981d03c1/examples/jsm/lines/LineMaterial.js#L35

https://github.com/mrdoob/three.js/blob/cf25a80737db2914f0b36e7c77af23a9981d03c1/examples/jsm/lights/RectAreaLightUniformsLib.js#L46

3. Again this all works when importing from the source files, these are blockers on the flat bundle only. I realize we're only supporting the flat bundle, but if the goal here is to allow people to modify them, wouldn't it make more sense to be working with the source files then?

4. Regardless, I think anyway we approach it we'll need to get rid of the `ShaderChunk`, `ShaderLib` and `UniformsLib` namespaces. I have this working locally with the shaders all self-contained in each of the `Materials`, and it's working great, the main blocker for me right now is supporting the "custom" shader use cases.

5. It's worth noting extending a material works as a way to override the shaders, I'm specifically talking about custom shaders that are built using `ShaderChunk`, `ShaderLib` and `UniformsLib` directly.

How would you guys feel about possibly opening-up a use case where the `src/*` files aren't just there for "rhetoric", but as a possible use case for custom shaders? Like instead of importing from `'three'` for a particular shader by name, you could just import from the source file? This way we could effectively remove all these namespaces from the flat bundle... 🤔

**Comment by pschroen** [comment]
Ya, @marcofugaro I've hit the same dead-end with `#include <...>` and `ShaderChunk`. I was able to fully remove `ShaderLib` and `UniformsLib`, but the entire shader system of three.js is built around these [includes](https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLProgram.js#L209).

And the savings from `ShaderLib` and `UniformsLib` alone are not enough to justify it. So super long story short, if we go any further it would be a huge breaking change now at this point. This library is really built to function around having all the shaders available in the flat bundle, and it also allows for custom shaders from the flat bundle if we leave it like that.

Well, I gave it my best shot guys, and it's been a good learning experience being knee-deep in the three.js renderer and shaders. There's not much we can do about the remaining static and `.prototype.*` properties either:

> I think it's more a problem of modernization, so either webpack needs to add tree-shaking support for .prototype.* properties, or we move to class fields, whichever comes first. 😅

After #24336 is merged I think we can close this issue now, we've gone as far as we can with the shaders and webpack for now. 🫠

**Comment by donmccurdy** [comment]
Is there a short summary of the issue with `#include <...>`? I can see it is a problem that the `#include` macros refer to a shared global namespace. Perhaps that could be replaced with namespaces specific to a material type?

```js
import { x } from 'path/to/chunks/x.glsl.js';
import { y } from 'path/to/chunks/y.glsl.js';

class MeshStandardMaterial extends Material { ... }

MeshStandardMaterial.ShaderChunk = { x, y };

export { MeshStandardMaterial };
```

Still a breaking change, but manageable. Or have I missed the issue?

**Comment by pschroen** [comment]
You got it, that would be the next step I was going to take but feeling it's not really worth it. See https://github.com/mrdoob/three.js/issues/21665#issuecomment-823930327

Writing custom shaders with three.js is already a little awkward, and I think this would make it even more complicated. It's more convenient having all the chunks in one place I think. 🙃

**Comment by gkjohnson** [comment]
> I can see it is a problem that the #include macros refer to a shared global namespace. Perhaps that could be replaced with namespaces specific to a material type?

My impression is that there are a few issues that could ultimately affect the end user if not handled properly. Take this all with a grain of salt but this is what I've gathered over from roughly following these conversations:

1. The main shader materials all rely on these shader chunk includes which reference a map of shader chunks by string id which aren't tree shakeable.
2. These could be removed in a variety of ways (inlining using templates, for example, removing #includes entirely) but ultimately end users have a lot of code that rely on pattern matching those include statements to augment built in shaders or use them in their own shader code.

Point 2 is what makes this hard because it means we can't just remove the `#include` statements outright or only just include the ones required for just the instantiated built-in materials. Not to add more noise to all this but there may be some approach with a minimal migration path for existing users that implicitly rely on some of these shader chunks. 

---

- Add a registration-style "ShaderChunkLib" that affords registering shader chunks and uniform sets.
- In the constructor of all the built-in materials each required shader chunk and uniform set could could be registered.
- Then users could register their own shader chunks / uniform sets (including built-in ones that aren't implicitly registered).
- A function like `registerAllChunks` could be included that existing users could run to enable all their existing shader code to continue to work until they've move to import just what's needed.

If I'm understanding all this tree shaking stuff then that should afford the ability for only the required shader chunks and uniform sets to be included in a tree-shaken bundle. The built-in material code might look like this:

```js
import { bsdfs, uv_pars_fragment, ... } from 'path/to/chunks/ShaderChunk.js';
import { ShaderChunkLib } from 'path/to/chunks/ShaderChunkLib.js';

class MeshStandardMaterial extends Material { 

  constructor() {

    super();
    ShaderChunkLib.register( 'bsdfs', bsdfs );
    ShaderChunkLib.register( 'uv_pars_fragment', uv_pars_fragment );
    // ...

  }

}

// ... migration function for end users

import { registerAllShaderChunks } from 'three';

registerAllShaderChunks();

// all built-in chunks available for user-written shaders
```

Admittedly the above also looks like a huge pain to maintain. But perhaps there's a more svelte pattern for this (like what @donmccurdy suggested). Though at the least it seems like something like this would get things working without dramatically impacting end users.

**Comment by pschroen** [comment]
Interesting, ya something that's evaluated at runtime would work much better with tree-shaking, to quote [agadoo](https://github.com/Rich-Harris/agadoo) again:

> Don't create instances of things on initial evaluation — instantiate lazily, when the functions you export are called

**Comment by LeviPesin** [comment]
One more related discussion:
https://github.com/mrdoob/three.js/pull/23368#issuecomment-1024741341

**Comment by pschroen** [comment]
As an update on [agadoo](https://github.com/rich-harris/agadoo), Rich Harris has updated it to the latest version of [Rollup](https://rollupjs.org/) with the release of `v3.0.0` (Rich-Harris/agadoo#16), and as a quick unit test:

```sh
git clone --depth=1 https://github.com/mrdoob/three.js.git
cd three.js
npx agadoo
```

**Comment by looeee** [comment]
`npx agadoo` fails, however if I remove these lines:

``` js
if (typeof __THREE_DEVTOOLS__ !== "undefined") {
	__THREE_DEVTOOLS__.dispatchEvent(
		new CustomEvent("register", {
			detail: {
				revision: REVISION,
			},
		})
	);
}

if (typeof window !== "undefined") {
	if (window.__THREE__) {
		console.warn("WARNING: Multiple instances of Three.js being imported.");
	} else {
		window.__THREE__ = REVISION;
	}
}

```

Then run `npm build` and `npx agadoo` again, it succeeds. So I guess these globals are the last remaining issue.

**Comment by Mugen87** [comment]
It seems the `three.js` dev tools are not maintained for quite a while anymore so we can probably remove the `__THREE_DEVTOOLS__` related code.

The multiple instance check is helpful in certain cases but I wonder if it's more important than full tree shake support. Granted, it's a bit of a hack that the library writes to `window`...

**Comment by donmccurdy** [comment]
> The multiple instance check is helpful in certain cases but I wonder if it's more important than full tree shake support.

Personally I think the multi-import check is more helpful, yes, if everything _except the 6-7 lines of code required are being tree-shaken. If dev tools were actively maintained I would vote to keep those lines too, but as it is... perhaps not. The difference between 99.9% tree-shaking and 100% tree-shaking is basically nothing.

Or it may be possible to wrap those lines in a block like this:

```javascript
if (process?.env?.NODE_ENV === 'development') {
  // ...
}
```

Not sure about agadoo, but bundlers should strip that code except during local development.

**Comment by marcofugaro** [comment]
> Or it may be possible to wrap those lines in a block like this:

Nope, bundlers do a string replacement of the `process.env.NODE_ENV` word with the string `"development"` (or `"production"`) at build time, so the example you posted doesn't work.

The multi-import check has been proven useful for me multiple times, so I vote we leave it.

**Comment by donmccurdy** [comment]
Probably possible to work around that?

```javascript
const process = typeof global === 'undefined' ? {env: {}} : global.process;

if (process.env.NODE_ENV === 'development') {
  // ...
}
```

But either way, I don't think having those few lines remaining is a problem.

**Comment by looeee** [comment]
> if everything except the 6-7 lines of code required are being tree-shaken

This is a good question, which of these is true?

1. The rest of the library is correctly tree-shaken and just these lines remain
2. Bundlers encounter these lines and bail out of tree-shaking entirely

If it's 1. then I don't see any problem with these lines remaining.

**Comment by marcofugaro** [comment]
> Bundlers encounter these lines and bail out of tree-shaking entirely

In my experience, that doesn't happen.

**Comment by mrdoob** [comment]
Should we close this then?

**Comment by pschroen** [comment]
Agreed, you want to do the honours, or should I close it? 😅

**Comment by mrdoob** [comment]
🎉
